fv3core/grid/gnomonic.py

```diff
@@ -0,0 +1,844 @@
+
+
+def fortran_vector_spherical_angle(e1, e2, e3):
+    """
+   The Fortran version
+    Given x/y/z tuples, compute the spherical angle between
+    them according to:
+!           p3
+!         /
+!        /
+!       p_center ---> angle
+!         \
+!          \
+!           p2
+    This angle will always be less than Pi.
+    """
+
+    # ! Vector P:
+    #    px = e1(2)*e2(3) - e1(3)*e2(2)
+    #    py = e1(3)*e2(1) - e1(1)*e2(3)
+    #    pz = e1(1)*e2(2) - e1(2)*e2(1)
+    # ! Vector Q:
+    #    qx = e1(2)*e3(3) - e1(3)*e3(2)
+    #    qy = e1(3)*e3(1) - e1(1)*e3(3)
+    #    qz = e1(1)*e3(2) - e1(2)*e3(1)
+
+    # Vector P:
+    px = e1[1] * e2[2] - e1[2] * e2[1]
+    py = e1[2] * e2[0] - e1[0] * e2[2]
+    pz = e1[0] * e2[1] - e1[1] * e2[0]
+    # Vector Q:
+    qx = e1[1] * e3[2] - e1[2] * e3[1]
+    qy = e1[2] * e3[0] - e1[0] * e3[2]
+    qz = e1[0] * e3[1] - e1[1] * e3[0]
+    ddd = (px * px + py * py + pz * pz) * (qx * qx + qy * qy + qz * qz)
+
+    if ddd <= 0.0:
+        angle = 0.0
+    else:
+        ddd = (px * qx + py * qy + pz * qz) / math.sqrt(ddd)
+        if abs(ddd) > 1.0:
+            # FIX (lmh) to correctly handle co-linear points (angle near pi or 0)
+            if ddd < 0.0:
+                angle = 4.0 * math.atan(1.0)  # should be pi
+            else:
+                angle = 0.0
+        else:
+            angle = math.acos(ddd)
+    return angle
+
+
+def spherical_angle(p_center, p2, p3, np):
+    """
+    Given ndarrays whose last dimension is x/y/z, compute the spherical angle between
+    them according to:
+!           p3
+!         /
+!        /
+!       p_center ---> angle
+!         \
+!          \
+!           p2
+    This angle will always be less than Pi.
+    """
+
+    # ! Vector P:
+    #    px = e1(2)*e2(3) - e1(3)*e2(2)
+    #    py = e1(3)*e2(1) - e1(1)*e2(3)
+    #    pz = e1(1)*e2(2) - e1(2)*e2(1)
+    # ! Vector Q:
+    #    qx = e1(2)*e3(3) - e1(3)*e3(2)
+    #    qy = e1(3)*e3(1) - e1(1)*e3(3)
+    #    qz = e1(1)*e3(2) - e1(2)*e3(1)
+
+    p = np.cross(p_center, p2)
+    q = np.cross(p_center, p3)
+    angle = np.arccos(
+        np.sum(p * q, axis=-1)
+        / np.sqrt(np.sum(p ** 2, axis=-1) * np.sum(q ** 2, axis=-1))
+    )
+    if not np.isscalar(angle):
+        angle[np.isnan(angle)] = 0.0
+    elif math.isnan(angle):
+        angle = 0.0
+
+    return angle
+
+
+#    ddd = (px*px+py*py+pz*pz)*(qx*qx+qy*qy+qz*qz)
+
+#    if ( ddd <= 0.0d0 ) then
+#         angle = 0.d0
+#    else
+#         ddd = (px*qx+py*qy+pz*qz) / sqrt(ddd)
+#         if ( abs(ddd)>1.d0) then
+#              angle = 2.d0*atan(1.0)    ! 0.5*pi
+#            !FIX (lmh) to correctly handle co-linear points (angle near pi or 0)
+#            if (ddd < 0.d0) then
+#               angle = 4.d0*atan(1.0d0) !should be pi
+#            else
+#               angle = 0.d0
+#            end if
+#         else
+#              angle = acos( ddd )
+#         endif
+#    endif
+
+#    spherical_angle = angle
```

Comment:
Remember to delete these before merging, if you like you can make reference to files and subroutine names in the Fortran code.